tests/unit/sharded_copying_manager_tests/common.py

```diff
@@ -0,0 +1,919 @@
+            self._test_state_cv.release()
+
+    def _test_state_cv_wait_with_timeout(self, start_time):
+        """Waits on the `_test_state_cv` condition variable, but will also throw an AssertionError if the wait
+        time exceeded the `start_time` plus `WAIT_TIMEOUT`.
+
+        @param start_time:  The start time when we first began waiting on this condition, in seconds past epoch.
+        @type start_time: Number
+        """
+        deadline = start_time + TestableCopyingManagerThreadedWorker.WAIT_TIMEOUT
+        self._test_state_cv.wait(timeout=(deadline - time.time()) + 0.5)
+        if time.time() > deadline:
+            raise AssertionError(
+                "Deadline exceeded while waiting on condition variable"
+            )
+
+
+class TestableCopyingManagerThreadedWorker(
+    CopyingManagerThreadedWorker, TestableCopyingManagerFlowController
+):
+    """An instrumented version of the CopyingManagerWorker which allows intercepting of requests sent, control when
+    the worker processes new logs, etc.
+
+    This allows for end-to-end testing of the core of the CopyingManagerWorker.
+
+    Doing this right is a bit complicated because the CopyingManagerWorker runs in its own thread.
+
+    Many members of this class are common with the TestableCopyingManager class
+    and mare oved to the base class - TestableCopyingManagerFlowController
+    """
+
+    __test__ = False
+
+    # The different points at which the CopyingManager can be stopped.  See below.
+    SLEEPING = "SLEEPING"
+    SENDING = "SENDING"
+    RESPONDING = "RESPONDING"
+
+    # To prevent tests from hanging indefinitely, wait a maximum amount of time before giving up on some test condition.
+    WAIT_TIMEOUT = 5.0
+
+    def __init__(self, configuration, api_key_config_entry, worker_id):
+        # Approach:  We will override key methods of CopyingManagerWorker, blocking them from returning until we tell
+        # to proceed.  This allows us to then do things like write new log lines while the CopyingManagerWorker is
+        # blocked.   Coordinating the communication between the two threads is done using one condition variable.
+        # We changed the CopyingManagerWorker to block in three places: while it is sleeping before it starts a new loop,
+        # when it invokes `_send_events` to send a new request, and when it blocks to receive the response.
+        # These three states are referred to as 'sleeping', 'sending', 'responding'.
+        #
+        # The CopyingManagerWorker will have state to record where it should next block (i.e., if it should block at
+        # 'sleeping' when it attempts to sleep).
+        #
+        # We can manipulate this state, notifying changes on the conditional variable
+        # by using following methods: 'wait_for_rpc', 'perform_scan', 'perform_pipeline_scan'
+        # The CopyingManagerWorker will block in this state (and indicate it is blocked) until the
+        # we set a new state to block at.
```

Comment:
This looks like the same comment as above in `TestableCopyingManagerFlowController`.  I would just say see those comments here.